src/Diag/Result.hs

```diff
@@ -146,6 +152,27 @@ data ErrCtx where
 instance Show ErrCtx where
   showsPrec p (ErrCtx c) = showParen (p > 10) $ showString "ErrCtx " . diagToShowS c
 
+-- | Some error help type. Right now, this just requries a ToDiagnostics instance for the type.
+data ErrHelp where
+  ErrHelp :: ToDiagnostic diag => diag -> ErrHelp
+
+instance Show ErrHelp where
+  showsPrec p (ErrHelp h) = showParen (p > 10) $ showString "ErrHelp " . diagToShowS h
+
+-- | Some error support type. Right now, this just requries a ToDiagnostics instance for the type.
+data ErrSupport where
+  ErrSupport :: ToDiagnostic diag => diag -> ErrSupport
+
+instance Show ErrSupport where
+  showsPrec p (ErrSupport s) = showParen (p > 10) $ showString "ErrSupport " . diagToShowS s
+
+-- | Some error doc type. Right now, this just requries a ToDiagnostics instance for the type.
+data ErrDoc where
+  ErrDoc :: ToDiagnostic diag => diag -> ErrDoc
+
+instance Show ErrDoc where
```

Comment:
Generally with Haskell, the expectation is that `read . show = id` which it does not in this case. I would let the compiler derive the `Show` instance and then use either `toString` or `toText` as appropriate.

Reply:
On second thought, I think it doesn't matter in this case. There's also no guarantee it will be derivable I think. Sorry!